mod-source-record-storage-server/src/main/java/org/folio/services/RecordServiceImpl.java

```diff
@@ -250,6 +275,54 @@ public Future<Void> updateRecordsState(String matchedId, RecordState state, Reco
     return recordDao.updateRecordsState(matchedId, state, recordType, tenantId);
   }
 
+  private Future<Record> setMatchedIdForRecord(Record record, String tenantId) {
+    String marcField999s = getFieldFromMarcRecord(record, TAG_999, INDICATOR, INDICATOR, SUBFIELD_S);
+    if (marcField999s != null) {
+      // Set matched id from 999$s marc field
+      LOG.debug(format("setMatchedIdForRecord:: Set matchedId: %s from 999$s field for record with id: %s", marcField999s, record.getId()));
+      return Future.succeededFuture(record.withMatchedId(marcField999s));
+    }
+    Promise<Record> promise = Promise.promise();
+    String externalId = RecordDaoUtil.getExternalId(record.getExternalIdsHolder(), record.getRecordType());
+    IdType idType = RecordDaoUtil.getExternalIdType(record.getRecordType());
+
+    if (externalId != null && idType != null && record.getState() == Record.State.ACTUAL) {
+      recordDao.getSourceRecordByExternalId(externalId, idType, RecordState.ACTUAL, tenantId)
+        .onComplete((ar) -> {
+          if (ar.succeeded()) {
+            Optional<SourceRecord> sourceRecord = ar.result();
+            if (sourceRecord.isPresent()) {
+              // Set matched id from existing source record
+              String sourceRecordId = sourceRecord.get().getRecordId();
+              LOG.debug(format("setMatchedIdForRecord:: Set matchedId: %s from source record for record with id: %s",
+                sourceRecordId, record.getId()));
+              promise.complete(record.withMatchedId(sourceRecordId));
+            } else {
+              // Set matched id same as record id
+              LOG.debug(format("setMatchedIdForRecord:: Set matchedId same as record id: %s", record.getId()));
+              promise.complete(record.withMatchedId(record.getId()));
+            }
+          } else {
+            LOG.warn("setMatchedIdForRecord:: Error while retrieving source record");
+            promise.fail(ar.cause());
+          }
+        });
+    } else {
+      // Set matched id same as record id
+      promise.complete(record.withMatchedId(record.getId()));
+    }
+    return promise.future().onSuccess(r -> addFieldToMarcRecord(r, TAG_999, SUBFIELD_S, r.getMatchedId()));
```

Comment:
@RomanChernetskyi  Is there an opportunity to divide this method via other methods? The body in this "if" condition seems so complex. Thank you in advance.

Reply:
Done, thanks